### PR TITLE
Add Python 3.12 to CI and reorganize tests

### PR DESCRIPTION
**Pull request by namurphy**
Python 3.12 came out in October 2023.  This PR adds it to our CI suite.  

It probably won't work because Numba hasn't been released for this version of Python.

### CODE REVIEW

**Comment by namurphy** [comment]
A release candidate for Numba for Python 3.12 just came out, so hopefully we can merge this soon!

File: requirements.txt [hunk 1]
```diff
@@ -96,7 +96,7 @@ future==0.18.3
     # via uncertainties
 h5py==3.10.0
     # via plasmapy (setup.py)
-hypothesis==6.97.3
+hypothesis==6.97.4
```
Comment (namurphy) [inline requirements.txt:99]:
```suggestion
hypothesis==6.97.3
```

File: requirements.txt [hunk 2]
```diff
@@ -113,7 +113,7 @@ iniconfig==2.0.0
     # via pytest
 ipykernel==6.29.0
     # via plasmapy (setup.py)
-ipython==8.20.0
+ipython==8.21.0
```
Comment (namurphy) [inline requirements.txt:116]:
```suggestion
ipython==8.20.0
```

File: requirements.txt [hunk 3]
```diff
@@ -264,7 +264,7 @@ pillow==10.2.0
     #   matplotlib
     #   plasmapy (setup.py)
     #   sphinx-gallery
-platformdirs==4.1.0
+platformdirs==4.2.0
```
Comment (namurphy) [inline requirements.txt:267]:
```suggestion
platformdirs==4.1.0
```

**Comment by namurphy** [comment]
Numba 0.59.0 was just released with Python 3.12 support!  This means that as of today, we can run PlasmaPy on Python 3.12.  

I'm still hoping eventually that we drop the dependency on Numba so that we don't have to wait a few months after a new Python release...

File: .github/workflows/weekly-tests.yml [hunk 4]
```diff
@@ -6,26 +6,6 @@ on:
...
-    steps:
-    - name: Set up conda
-      uses: conda-incubator/setup-miniconda@v3
-      with:
-        auto-update-conda: true
-        python-version: 3.9
-
-    - name: Install PlasmaPy
-      run: conda install -c ${{ matrix.channel }} --yes plasmapy
-
```
Comment (namurphy) [inline .github/workflows/weekly-tests.yml:28]: This is moved below for organizational purposes.

**Comment by namurphy** [comment]
@ejohnson-96 — It lists a few checks as *Expected — Waiting for status to be reported* because I renamed the tests in this pull request, and haven't yet updated the list of required tests in the settings of the repository.

**Review by ejohnson-96 — APPROVED** [review]
👍

**Pull request merged**